File: q0445/q0445.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
    {
        stack<int> s1;
        stack<int> s2;
        while (l1 != nullptr)
        {
            s1.push(l1->val);
            l1 = l1->next;
        }

        while (l2 != nullptr)
        {
            s2.push(l2->val);
            l2 = l2->next;
        }

        int currSum = 0;
        int carry = 0;
        ListNode *sumNode = new ListNode();
        while (!s1.empty() || !s2.empty())
        {
            if (!s1.empty())
            {
                currSum += s1.top();
                s1.pop();
            }

            if (!s2.empty())
            {
                currSum += s2.top();
                s2.pop();
            }

            sumNode->val = currSum % 10;
            carry = currSum / 10;

            ListNode *newSumNode = new ListNode(carry, sumNode);
            sumNode = newSumNode;

            currSum = carry;
        }

        return sumNode->val == 0 ? sumNode->next : sumNode;
    }
};

File: q0445/q0445.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        Stack<Integer> s1 = new Stack<>();
        Stack<Integer> s2 = new Stack<>();

        while (l1 != null) {
            s1.push(l1.val);
            l1 = l1.next;
        }

        while (l2 != null) {
            s2.push(l2.val);
            l2 = l2.next;
        }

        int currSum = 0;
        int carry = 0;
        ListNode sumHead = new ListNode();
        while (!s1.isEmpty() || !s2.isEmpty()) {
            int n1 = s1.isEmpty() ? 0 : s1.pop();
            int n2 = s2.isEmpty() ? 0 : s2.pop();
            currSum += n1 + n2;

            sumHead.val = currSum % 10;
            carry = currSum / 10;

            ListNode newSumHead = new ListNode(carry, sumHead);
            sumHead = newSumHead;

            currSum = carry;
        }
        return sumHead.val == 0 ? sumHead.next : sumHead;
    }
}